src/operators/join/hash_join.cc

```diff
@@ -0,0 +1,325 @@
+    // TODO: for now, we assume the join column is INT64 type.
+    auto chunk = std::static_pointer_cast<arrow::Int64Array>(col->chunk(i));
+    const uint8_t *filter_data = nullptr;
+    if (filter != nullptr) {
+      filter_data = filter->chunk(i)->data()->GetValues<uint8_t>(1, 0);
+    }
+    for (std::uint32_t row = 0; row < chunk->length(); row++) {
+      if (filter_data == nullptr || arrow::BitUtil::GetBit(filter_data, row)) {
+        auto record_id_itr = hash_table_->find(chunk->Value(row));
+        if (record_id_itr != hash_table_->end()) {
+          auto record_ids = record_id_itr->second;
+          record_ids.push_back({(uint32_t)chunk_offsets[i] + row, (uint16_t)i});
+          (*hash_table_)[chunk->Value(row)] = record_ids;
+        } else {
+          (*hash_table_)[chunk->Value(row)] = {
+              {(uint32_t)chunk_offsets[i] + row, (uint16_t)i}};
+        }
+      }
+    }
+  }
+}
+
+void HashJoin::ProbeHashTableBlock(
+    const std::shared_ptr<arrow::ChunkedArray> &probe_col,
+    const std::shared_ptr<arrow::ChunkedArray> &probe_filter, int batch_idx,
+    int batch_size, std::vector<uint64_t> chunk_row_offsets) {
+  int base_idx = batch_idx * batch_size;
+  auto hash_table_end = hash_table_->end();
+  for (size_t i = base_idx;
+       (i < base_idx + batch_size && i < probe_col->num_chunks()); ++i) {
+    int num_indices = 0;
+
+    auto chunk = probe_col->chunk(i);
+    auto left_data = chunk->data()->GetValues<uint64_t>(1, 0);
+    const uint8_t *filter = nullptr;
+
+    if (probe_filter != nullptr) {
+      auto filter_chunk = probe_filter->chunk(i);
+      filter = (const uint8_t *)filter_chunk->data()->GetValues<uint8_t>(1, 0);
+    }
+
+    // The indices of the rows joined in chunk i
+    int64_t chunk_len = chunk->length();
+    uint32_t *l_idxs = (uint32_t *)malloc(sizeof(uint32_t) * chunk_len);
+    uint32_t *r_idxs = (uint32_t *)malloc(sizeof(uint32_t) * chunk_len);
+    auto idx_len = chunk_len;
+    auto offset = chunk_row_offsets[i];
+    for (auto row = 0; row < chunk_len; row++) {
+
+      if (filter == nullptr || arrow::BitUtil::GetBit(filter, row)) {
+        auto record_itr = hash_table_->find(left_data[row]);
+
+        if (record_itr != hash_table_end) {
+          for (auto record_id : record_itr->second) {
+            if (idx_len <= num_indices) {
+              idx_len <<= 1;
+              l_idxs = (uint32_t *)realloc(l_idxs, sizeof(uint32_t) * idx_len);
+              r_idxs = (uint32_t *)realloc(r_idxs, sizeof(uint32_t) * idx_len);
+            }
+            l_idxs[num_indices] = row + offset;
+            r_idxs[num_indices] = record_id.index;
+            ++num_indices;
+          }
+        }
+      }
+    }
+
+    left_indices_[i] = std::vector<uint32_t>(l_idxs, l_idxs + num_indices);
+    right_indices_[i] = std::vector<uint32_t>(r_idxs, r_idxs + num_indices);
+
+    free(l_idxs);
+    free(r_idxs);
+  }
+}
+
+void HashJoin::ProbeHashTable(
```

Comment:
Shall we later separate hash join into two phase (as two operators) - hash build + hash probe?

Reply:
Sure, will do them in the coming work